<?xml version="1.0" encoding="utf-8"?>
<!-- This Xacro file was converted from a URDF file -->
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="omnifinal">

  <!-- Include the Gazebo-specific configurations -->
  <xacro:include filename="$(find omnifinal)/urdf/omnifinal.gazebo" />

  <!-- Define properties for reuse -->
  <xacro:property name="scale" value="1.0"/>

  <!-- root link, on the ground just below the model origin -->
  <link name="base_footprint"></link>

	<joint name="base_footprint_joint" type="fixed">
		<origin xyz="0 0 0" rpy="0 0 0" />
		<parent link="base_footprint" />
		<child link="base_link" />
	</joint>

  <link
    name="base_link">
    <inertial>
      <origin
        xyz="0.00211440695779285 -0.000172345786933302 0.0214923159791835"
        rpy="0 0 0" />
      <mass
        value="0.213265827988128" />
      <inertia
        ixx="0.000459624639615948"
        ixy="-6.64436047700312E-08"
        ixz="4.28549936785738E-06"
        iyy="0.00096819045435354"
        iyz="1.42696555292235E-07"
        izz="0.00127034015073323" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/base_link.STL" />
      </geometry>
      <material
        name="gray">
        <color
          rgba="0.792156862745098 0.819607843137255 0.933333333333333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/base_link.STL" />
      </geometry>
    </collision>
  </link>
  <link
    name="fl_link">
    <inertial>
      <origin
        xyz="-1.7347E-16 0.018355 7.1124E-17"
        rpy="0 0 0" />
      <mass
        value="0.04956" />
      <inertia
        ixx="5.248E-06"
        ixy="-6.9309E-19"
        ixz="-5.4019E-19"
        iyy="6.394E-06"
        iyz="5.9917E-19"
        izz="5.248E-06" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/fl_link.STL" />
      </geometry>
      <material
        name="black">
        <color
          rgba="0 0 0 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/fl_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="fl_joint"
    type="continuous">
    <origin
      xyz="0.04 0.0473999999999994 0.0085"
      rpy="0 0 0" />
    <parent
      link="base_link" />
    <child
      link="fl_link" />
    <axis
      xyz="0 1 0" />
  </joint>
  <link
    name="fr_link">
    <inertial>
      <origin
        xyz="8.3267E-17 -0.017845 1.6306E-16"
        rpy="0 0 0" />
      <mass
        value="0.04956" />
      <inertia
        ixx="5.248E-06"
        ixy="3.8308E-18"
        ixz="7.3101E-19"
        iyy="6.394E-06"
        iyz="-8.4723E-20"
        izz="5.248E-06" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/fr_link.STL" />
      </geometry>
      <material
        name="black">
        <color
          rgba="0 0 0 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/fr_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="fr_joint"
    type="continuous">
    <origin
      xyz="0.04 -0.0473999999999993 0.00850000000000001"
      rpy="0 0 0" />
    <parent
      link="base_link" />
    <child
      link="fr_link" />
    <axis
      xyz="0 1 0" />
  </joint>
  <link
    name="bl_link">
    <inertial>
      <origin
        xyz="-3.6776E-16 0.017845 1.5266E-16"
        rpy="0 0 0" />
      <mass
        value="0.04956" />
      <inertia
        ixx="5.248E-06"
        ixy="3.0199E-18"
        ixz="-7.3136E-19"
        iyy="6.394E-06"
        iyz="2.6246E-18"
        izz="5.248E-06" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/bl_link.STL" />
      </geometry>
      <material
        name="black">
        <color
          rgba="0 0 0 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/bl_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="bl_joint"
    type="continuous">
    <origin
      xyz="-0.04 0.0473999999999994 0.0085"
      rpy="0 0 0" />
    <parent
      link="base_link" />
    <child
      link="bl_link" />
    <axis
      xyz="0 1 0" />
  </joint>
  <link
    name="br_link">
    <inertial>
      <origin
        xyz="-6.0368E-16 -0.017845 2.0123E-16"
        rpy="0 0 0" />
      <mass
        value="0.04956" />
      <inertia
        ixx="5.248E-06"
        ixy="3.4778E-18"
        ixz="-7.6167E-19"
        iyy="6.394E-06"
        iyz="5.7911E-18"
        izz="5.248E-06" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/br_link.STL" />
      </geometry>
      <material
        name="black">
        <color
          rgba="0 0 0 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/br_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="br_joint"
    type="continuous">
    <origin
      xyz="-0.04 -0.0474 0.0085"
      rpy="0 0 0" />
    <parent
      link="base_link" />
    <child
      link="br_link" />
    <axis
      xyz="0 1 0" />
  </joint>
  <link
    name="arm_link1">
    <inertial>
      <origin
        xyz="0.019999 7.2856E-18 0.00016695"
        rpy="0 0 0" />
      <mass
        value="0.0045341" />
      <inertia
        ixx="7.5829E-08"
        ixy="-4.3104E-22"
        ixz="-6.5134E-09"
        iyy="8.5628E-07"
        iyz="-9.1565E-24"
        izz="8.5602E-07" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/arm_link1.STL" />
      </geometry>
      <material
        name="gray">
        <color
          rgba="0.5 0.5 0.5 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/arm_link1.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="arm_joint1"
    type="revolute">
    <origin
      xyz="0.092 0 0.036"
      rpy="0 0 0" />
    <parent
      link="base_link" />
    <child
      link="arm_link1" />
    <axis
      xyz="0 1 0" />
    <limit
      lower="-0.53"
      upper="0.53"
      effort="0.3"
      velocity="0.2" />
  </joint>
  <link
    name="arm_link2">
    <inertial>
      <origin
        xyz="2.984E-05 2.3489E-17 -0.021042"
        rpy="0 0 0" />
      <mass
        value="0.005267" />
      <inertia
        ixx="1.1161E-06"
        ixy="5.931E-24"
        ixz="-3.6056E-11"
        iyy="1.0529E-06"
        iyz="5.9347E-22"
        izz="1.5124E-07" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/arm_link2.STL" />
      </geometry>
      <material
        name="gray">
        <color
          rgba="0.5 0.5 0.5 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://omnifinal/meshes/arm_link2.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="arm_joint2"
    type="revolute">
    <origin
      xyz="0.039968 0 0.001599"
      rpy="0 0 0" />
    <parent
      link="arm_link1" />
    <child
      link="arm_link2" />
    <axis
      xyz="0 1 0" />
    <limit
      lower="-1.57"
      upper="0"
      effort="0.3"
      velocity="0.2" />
  </joint>

  <!-- Transmission for arm_joint1 -->
  <transmission name="arm_joint1_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="arm_joint1">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="arm_joint1_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <!-- Transmission for arm_joint2 -->
  <transmission name="arm_joint2_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="arm_joint2">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="arm_joint2_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <!-- Camera joint -->
  <joint name="camera_joint" type="fixed">
    <origin xyz="0.05 0 0.055" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="camera_link"/>
  </joint>

  <!-- Camera link -->
  <link name="camera_link">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.008 0.008 0.008"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.008 0.008 0.008"/>
      </geometry>
      <material name="red">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>
    <inertial>
      <mass value="1e-5"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>

	<!-- rplidar Laser joint -->
  <joint name="rplidar_joint" type="fixed">
		<axis xyz="0 1 0" />
		<origin xyz="-0.02 0 0.06" rpy="0 0 0"/>
		<parent link="base_link"/>
		<child link="laser"/>
	</joint>

	<!-- rplidar Laser link -->
	<link name="laser">
		<collision>
			<origin xyz="-0.1 0 0.055" rpy="1.5707 0 3.1412"/>
			<geometry>
				<mesh filename="package://omnifinal/meshes/rplidar.dae" scale="0.001 0.001 0.001" />
			</geometry>
		</collision>

		<visual>
			<origin xyz="0 0 0" rpy="1.5707 0 4.71"/>
			<geometry>
				<mesh filename="package://omnifinal/meshes/rplidar.dae" scale="0.001 0.001 0.001" />
			</geometry>
      <material name="black"/>
		</visual>

		<inertial>
			<mass value="${1e-6*scale}" />
			<origin xyz="0 0 0.058" rpy="1.5707 0 4.71"/>
			<inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0" />
		</inertial>
	</link>


  <!-- IMU -->
	<joint name="imu_sensor_joint" type="continuous">
        <axis xyz="0 0 1" />
        <origin xyz="0 0 0.015" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="imu_link_ned"/>
    </joint>
     
    <link name="imu_link_ned">
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <box size="0.02 0.02 0.02"/>
            </geometry>
        </collision>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <box size="0.02 0.02 0.02"/>
            </geometry>
            <material name="red">
                <color rgba="1 0 0 1"/>
            </material>
        </visual>
        <inertial>
            <mass value="${0.001*scale}" />
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <inertia ixx="0.00000015" ixy="0" ixz="0" iyy="0.00000015" iyz="0" izz="0.00000015" />
        </inertial>
    </link>
</robot>